nsqd/nsqd.go

```diff
@@ -80,6 +81,26 @@ type NSQD struct {
 func New(opts *Options) (*NSQD, error) {
 	var err error
 
+	if opts.BroadcastHTTPPort == 0 {
+		_, portStr, err := net.SplitHostPort(opts.HTTPAddress)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse HTTP address (%s) - %s", opts.HTTPAddress, err)
+		}
+
+		portInt, _ := strconv.ParseInt(portStr, 10, 32)
+		opts.BroadcastHTTPPort = int(portInt)
```

Comment:
The reason why all the tests are failing is because they use port `0` to listen on a random free port.

E.g. https://github.com/nsqio/nsq/pull/1297/files#diff-6b512c5c86dadb7cbbe6f77d08d7a080d3b0d8780eb8801ee41dc59c803d9c2eL19 lazily uses `RealTCPAddr`() to send the actual port that was bound.

Reply:
... so if this moves down below where the Listeners are created, and uses `RealTcpAddr()` like before, that could work ...